### PR TITLE
Added Binary Focal Cross Entropy

### PR DESCRIPTION
**Pull request by Void426**
### Describe your change:
Added Binary Focal Cross Entropy loss function. It's a variation of binary cross-entropy that addresses class imbalance by 
focusing on hard examples.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by Void426** [comment]
@tianyizheng02 Could you please review this?

**Review by tianyizheng02 — CHANGES_REQUESTED** [review]
All loss function files were consolidated into `machine_learning/loss_functions.py` in #10737. Could you move your new code into that file?

**Comment by Void426** [comment]
> All loss function files were consolidated into `machine_learning/loss_functions.py` in #10737. Could you move your new code into that file?

Done.

File: machine_learning/loss_functions.py [hunk 1]
```diff
@@ -39,6 +39,59 @@ def binary_cross_entropy(
...
+    y_pred = np.clip(y_pred, epsilon, 1 - epsilon)
+
+    # Focal loss calculation
+    bcfe_loss = -(
+        alpha * (1 - y_pred) ** gamma * y_true * np.log(y_pred)
+        + (1 - alpha) * y_pred**gamma * (1 - y_true) * np.log(1 - y_pred)
+    )
+
+    # Take the mean over all samples
+    return np.mean(bcfe_loss)
```
Comment (tianyizheng02) [inline machine_learning/loss_functions.py:92]:
```suggestion
    # Clip predicted probabilities to avoid log(0)
    y_pred = np.clip(y_pred, epsilon, 1 - epsilon)

    bcfe_loss = -(
        alpha * (1 - y_pred) ** gamma * y_true * np.log(y_pred)
        + (1 - alpha) * y_pred**gamma * (1 - y_true) * np.log(1 - y_pred)
    )

    return np.mean(bcfe_loss)
```
For the first comment, there's nothing wrong with computing log(1), since that just evaluates to 0 with no issue. The issue is that the formula includes log(1 - x), which still may evaluate to log(0).

For the second and third comments, I feel that the code and variable names were already self-explanatory and didn't need the comments.

**Review by tianyizheng02 — APPROVED** [review]

**Comment by tianyizheng02** [comment]
Thanks for your contribution!
